Initial draft for typed constants

rhuss opened the pull request:
Also renamed Error --> VerificationError

rhuss committed:
Initial draft for typed constants

Also renamed Error --> VerificationError

rhuss commented:
10:26  ro14nd I just did an initial suggestion for some verifier changes: https://github.com/rhuss/camel/pull/1
10:27  ro14nd here is the full interface --> https://github.com/rhuss/camel/blob/fa2156b349c031d9b82a3fe7148eb8510ec48389/camel-core/src/main/java/org/apache/camel/ComponentVerifier.java
10:27  ro14nd I renamed Error --> VerificationError
10:27  ro14nd and added a Code and Attribute interface to VerificationError
10:28  ro14nd StandardCode and StandardAttribute are enums implementing this interface.
10:28  ro14nd but it can be also easily customized by implementing the interface itself.
10:29  ro14nd there are helpers to create such code and attributes from strings: https://github.com/rhuss/camel/blob/fa2156b349c031d9b82a3fe7148eb8510ec48389/camel-core/src/main/java/org/apache/camel/impl/verifier/ResultErrorBuilder.java#L181
10:30  ro14nd so its typesafe but still flexible.
10:30  ro14nd drawback: its a bit more verbose

rhuss commented:
Other renamings:

* `Error.getParameters()` --> `getParameterKeys`
* `Error.getAttributes()` --> `getDetails()` (as the "Attribute" here is used as the key of the map.

rhuss commented:
@lburgazzoli : What do you think ? Is this going in the right direction ? Or is it over the top ?

Still open: 

* Where to define the attribute constant string values ?
* Do we want custom key in the serialized output or is the mapping enum name == string constant ok ? We could also go for lowercase enums.

rhuss commented:
I adapted only camel-core, for the initial draft.

lburgazzoli commented on camel-core/src/main/java/org/apache/camel/ComponentVerifier.java:
```diff
@@ -27,71 +27,91 @@
-                if (ObjectHelper.equal(scope, VALUES[i].name(), true)) {
-                    return VALUES[i];
+        public static Scope fromString(String scopeS) {
+            for (Scope scope : values()) {
```
Not a blocking issue but I would like to keep a static reference for the scopes if possible as Enum.values() generates a new array for each call.

rhuss replied:
ah, didn't know this. Will re-insert it.

lburgazzoli replied:
It may not hurt too much but it all depends on how clever the jvm is

lburgazzoli commented on camel-core/src/main/java/org/apache/camel/ComponentVerifier.java:
```diff
@@ -27,71 +27,91 @@
-            for (int i = 0; i < VALUES.length; i++) {
-                if (ObjectHelper.equal(scope, VALUES[i].name(), true)) {
-                    return VALUES[i];
+        public static Scope fromString(String scopeS) {
```
scopeS is a little obscure, maybe scopeName ?

rhuss replied:
I will revert the whole method back. Just wondered about the private static `VALUES`

lburgazzoli commented on camel-core/src/main/java/org/apache/camel/impl/verifier/DefaultComponentVerifier.java:
```diff
@@ -123,7 +123,7 @@ protected void verifyParametersAgainstCatalog(ResultBuilder builder, Map<String,
                 .map(entry ->
                     ResultErrorBuilder.withIllegalOption(entry.getKey(), entry.getValue())
-                        .attribute("enum.values", result.getEnumChoices(entry.getKey()))
+                        .detail(ResultErrorBuilder.errorAttribute("enum.values"), result.getEnumChoices(entry.getKey()))
```
Maybe we can have a ResultErrorBuilder.detail(string, ...) ? so we can make it a little less verbose

rhuss replied:
It's because I think of the 'dedail' == 'attribute'  + value So this method only created the key, not the full detail.

Maybe only 'attribute' ? (as 'error' can be already inferred from the context / class)

lburgazzoli commented:
Looks good to me, I've added a few minor comments

rhuss committed:
Update according to PR comments:

- use simpler names for the ResultErrorBuilder.code() and .attribute()
- restored Scope VALUES cache

rhuss commented:
I updated the PR. The next step would be to update the usage of this verifier in all the other camel components. 

If there are no objections, I will do this and open a PR against apache/camel repo

lburgazzoli commented on camel-core/src/main/java/org/apache/camel/impl/verifier/DefaultComponentVerifier.java:
```diff
@@ -123,7 +123,7 @@ protected void verifyParametersAgainstCatalog(ResultBuilder builder, Map<String,
                 .map(entry ->
                     ResultErrorBuilder.withIllegalOption(entry.getKey(), entry.getValue())
-                        .attribute("enum.values", result.getEnumChoices(entry.getKey()))
+                        .detail(ResultErrorBuilder.attribute("enum.values"), result.getEnumChoices(entry.getKey()))
```
thinking aloud, could it also make sense to have:

```java
public ResultErrorBuilder detail(String key, Object value) {
    return detail(ResultErrorBuilder.attribute("enum.values"), value);
}
```

rhuss replied:
yeah, indeed. let me add this.

rhuss replied:
has to rename it to 'asAttribute()' and 'asCode()' because code() conflicts with the regular builder method.

Looks now much nicer :)

rhuss committed:
Update and add builder for the ResultErrorBuilder

lburgazzoli commented:
lgtm :+1:

rhuss commented on camel-core/src/main/java/org/apache/camel/impl/verifier/DefaultComponentVerifier.java:
```diff
@@ -123,7 +123,7 @@ protected void verifyParametersAgainstCatalog(ResultBuilder builder, Map<String,
                 .map(entry ->
                     ResultErrorBuilder.withIllegalOption(entry.getKey(), entry.getValue())
-                        .detail(ResultErrorBuilder.attribute("enum.values"), result.getEnumChoices(entry.getKey()))
+                        .detail("enum.values", result.getEnumChoices(entry.getKey()))
```
I wonder whether we should use an upper case constant here, too. (and underscore instead of point).

Alternative I we could tune the enums to reflect nicer names. E.g. to lower, _ -> ., adding  a prefix for all standard enums.

However in this case we then can't use name() in the interface but would have to use toString() ( name() is not overwritable, but toString()).

Still we have the benefit to compare via '==' (to check whether its a specific enum) and 'instanceof' (to check whether its a standard enum).

@lburgazzoli wdyt ?

E.g. StandardCode.ILLEGAL_OPTION could translate then to 'std.illegal.option' (or 'def' as prefix ? or no prefix ?)

The drawback would be, that with JSON serialization (as we use), still the name would be used. Also its a bit harder to map.

However, if we can live with '_' instead of '.' then we could make all enums lowercase too (e.g. std_illegal_option)

rhuss replied:
on the other hand, if we leave it like this than you can easily see what is a standard option (without prefix) because its all uppercase. all non uppercase are custom values.

Only hurts the eye, with mixed case conventions ;-)

lburgazzoli replied:
enum.values can be part of StandardAttribute imho as it something you can get form standard validation against camel catalog so it is cross component and I do not think there is any issue with lower case enums (test it with mvn install -Psourcecheck) but I would keep it uppercase as it is visually clear that it is a constant/enum (at least for me). 

Then about the wire representation, I like the "std.x.y" style so if you have a way to tweak the json serialization stuffs go ahead :)

rhuss replied:
yeah, thats a bit tricky. But having mixed cases (like here with 'enum.values') is no problem ?

lower case enums are indeed a bit cumbersome. at least my IDEA show them still in italics ;-)

Let me adapt the rest of camel and open a wider discussion on the 'official' PR

lburgazzoli replied:
Maybe the easier solution is to keep all the standard stuffs uppercase with underscore (as it reflects most of the use cases in camel) and then you can eventually translate it on your side so we do not impose any "wire format" in camel

rhuss replied:
for me, upper case on the wire is fine, too. We will soon to start reading binary with HTTP/2 soon anyway ;-P At least for the ipaas-verifier service. Facing to the UI we can make it nicer, but then again, I guess people don't care that much.

rhuss committed:
Worked on and added documentation
